Perceptual hash

bjlittle opened the pull request:
## The Why...

<details>
<summary> Some background context... </summary>

Please refer to issue https://github.com/matplotlib/pytest-mpl/issues/146.

This PR extends the capability of `pytest-mpl` to support perceptual image hashes.

In a nutshell, the use of `cryptographic` hashing algorithms e.g., `sha256` et al, only support strict binary image identity. By their very nature, they are not able to offer any meaningful way of determining the "similarity" of non-identical images. As a result, they offer limited use when comparing images, as 1 single pixel makes a huge difference.

In the same guise, the `rms` + `tolerance` approach can also be brittle, albeit certainly better than using a `cryptographic` hash. Distilling the richness of the detail and structure inherent within an image down to a single valued statistic results in an incredibly lossy metric, and hence can also lead to a fragile and sensitive workflow.

The real world up-shot of all this woe is a tangible drain on precious developer resource/effort/time in order to check graphic test failures... and TBH we've all got better things to do with our day :wink: 

Over on [SciTools/iris](https://github.com/SciTools/iris)  we've been using perceptual image hashes for a few years now, and it's pretty much resolved the pain and fear induced by the `sha` and `rms` approaches. It's way more robust, and through controlling the `hash-size` and `high frequency factor` it's allowed us to find a comfortable and reasonably reliable sweet-spot for graphical testing.

Therefore, we're really keen to make perceptual hashing available in `pytest-mpl`, so that the greater `matplotlib` community can benefit (as well as ourselves).

</details>

## The What...

<details>
<summary> A high-level summary... </summary>
This PR is two-fold:

- it introduces a simple and hopefully light-weight framework that supports easily configurable hashing algorithms or `kernels` that can be used to create associated hashing libraries
- provisions the default `sha256` cryptographic hashing `kernel` (which `pytest-mpl` currently implements), but also introduces the `phash` perceptual hashing `kernel` (non-cryptographic) provisioned by https://github.com/JohannesBuchner/imagehash

I've attempted to keep the PR as minimal as possible, by making my changes as additive as possible. However, I have rationalised some sections of the code-base (hopefully nothing too controversial) and applied some minor consistency changes (which were only due to fresh eyes on unfamiliar code - so happy to be shot down btw).

This PR is currently a **work-in-progress**, so all feedback is welcomed :+1: 

I'm super keen for `pytest-mpl` devs to get a feel for what I'm proposing early on, so that we can shape it appropriately. But from a purely selfish perspective, I'd love to see this feature banked, as it potentially opens the door to extensible and more robust image testing.

In terms of work still TODO, I figure that I still require to (at least) do the following:

- [ ] Testing
  - [x] Fix everything I've broken :rofl: 
    - see Issue #159
    - [ ] reinstate `subtests` (require PR #163)
  - [x] Add extra test coverage
    - [x] test `KernelPHash`
    - [x] test `KernelSHA256`
    - [x] extend coverage in `test_pytest_mpl` for `phash`
    - [ ] summary coverage 
- [ ] refactor after PR #160 
- [ ] Pull-thru changes and/or check report summary generation:
  - [x] `json`
  - [ ] `html`
  - [ ] `basic-html`
- [ ] Documentation
  - [x] Check doc-strings
  - [ ] Update `README.rst`

BTW If you guys buy-into this PR, then I may need help or advise along the way to service the above TODOs. However, I'd like to try my best to address them myself (as a learning opportunity) but I'll quickly shout if I get stuck :wink:      
</details>

## API Changes...

<details>
<summary> What's New... </summary>
In terms of public API, I've extended it as follows :

- CLI options:
  - `--mpl-kernel` : `str`, either of [`sha256`, `phash`], default is `sha256` to maintain existing default hashing behaviour
  - `--mpl-hash-size` : `int`, default is `16` bits, giving a hash size of 16**2 = 256 bits, only applies to `phash`
  - `--mpl-hamming-tolerance` : `int`, default is `2` bits, only applies to `phash`
  - `--mpl-high-freq-factor` : `int`, default is `4`, only applies to `phash`

- INI options:
  - `mpl-kernel` : `str`
  - `mpl-hash-size` : `int`
  - `mpl-hamming-tolerance` : `int`
  - `mpl-high-freq-factor` : `int`

- `mpl_image_compare` marker kwargs:
  - `hamming_tolerance` : `int`

Note that, my choice of defaults is heavily bias towards what we settled on for `SciTools/iris`.

https://github.com/JohannesBuchner/imagehash offers several other different hashing algorithms, but I've only opted to implement support for `imagehash.phash`. Given the initial `kernel` framework in this PR, the potential is there to extend support for any of those hashing algorithms, or indeed any other novel approach from the community... well, that's the grand hope and intention here.

</details>

Closes https://github.com/matplotlib/pytest-mpl/issues/146

bjlittle commented on pytest_mpl/plugin.py:
```diff
@@ -44,26 +43,59 @@
-  Actual shape: {actual_shape}
-    {actual_path}"""
+#: Supported INET protocols.
+PROTOCOLS = ('http://', 'https://')
```
Decided to make some of the embedded defaults buried away in the code base more explicit... hope that's okay?

bjlittle commented on pytest_mpl/plugin.py:
```diff
@@ -44,26 +43,59 @@
+  Result Shape:
+    {result_shape}
+  Result Image:
+    {result_image}"""
```
Upgraded this to an error rather than a failure... the distinction seems subtly important.

bjlittle commented on pytest_mpl/plugin.py:
```diff
@@ -44,26 +43,59 @@
+  Result Image:
+    {actual}
+  Difference Image:
+    {diff}"""
```
Applied the `SHAPE_MISMATCH` template approach here also, which for me, really makes the code readable where it's used.

bjlittle commented on pytest_mpl/plugin.py:
```diff
@@ -92,48 +124,80 @@ def pytest_report_header(config, startdir):
+    msg = 'Hamming distance bit tolerance for similar image hashes.'
+    option = 'mpl-hamming-tolerance'
+    group.addoption(f'--{option}', help=msg, action='store')
+    parser.addini(option, help=msg)
```
Just tidied this section up a wee bit since I was in the space... hope that's not too controversial?

bjlittle commented on pytest_mpl/plugin.py:
```diff
@@ -521,32 +618,57 @@ def compare_image_to_hash_library(self, item, fig, result_dir, summary=None):
+                        f'{kernel_name!r} does not match configured runtime '
+                        f'kernel {self.kernel.name!r}.')
+                pytest.fail(emsg)
+
```
It's important to have the name of the `kernel` that created the hash library as metadata within the hash library.

This will avoid situations where the configured runtime `kernel` is different to the `kernel` that generated the hashes in the library.

Note that, the choice of `META_HASH_LIBRARY_KERNEL` i.e., `pytest-mpl-kernel`, intentionally can't parse as a Python variable, and thus means that it's safe to inject this key into the dictionary, as it's simply not possible for a user of `pytest-mpl` to decorate a test with such a name... well, that's my thinking anyways :wink:

ConorMacBride replied:
Should we also record the high frequency factor used to generate the hash library? (We can probably determine the hash size used from the hashes themselves.) That way, similar to what I was saying above about the kernel name, we can dynamically set the default hash size and high frequency factor to the values used when generating the hash library (only if these have not been explicitly configured by the user as the CLI or ini option). 

What do you think of this approach of getting the default hash settings from the hash library?

bjlittle replied:
Done. See https://github.com/matplotlib/pytest-mpl/pull/150/commits/5958c5157a2b583c412132a3568510d1ccee2e6b :+1:

bjlittle commented on pytest_mpl/kernels.py:
```diff
@@ -0,0 +1,223 @@
+
+    def __init__(self, plugin):
+        # Containment of the plugin allows the kernel to cherry-pick required state.
+        self._plugin = plugin
```
It's a weak contract, but the deal here is that a `kernel` is forbidden to alter the contents of the `self._plugin` - that's just looking for trouble.

So this is a read-only private property. I didn't want to over-engineer that contract to protect ourselves from ourselves, as I didn't see that as worthwhile... but you may think otherwise.

Cadair commented:
Thanks for the PR @bjlittle it looks great. It will probably be next week before I have time to go through it properly.

bjlittle commented:
@Cadair Okay, so I appear to have opened a big old can of worms testing-wise, by:

- updating the `summary` dictionary to contain the three new necessary entries for `kernel`, `hamming_distance` and `hamming_tolerance`
- tweaking and changing various status and summary messages

I guess at this point it would be good to know whether my suggested changes for the status messages et al are acceptable. 

Based on accommodating any review request alterations from yourself, I'm guessing that I then need to recreate/respin the expected `tests.subtests.summaries.*.json` files used by the `tests.subtests.test_subtests`, which mostly all seem pretty upset with me atm :wink:  

I may need some guidance on how to go about doing that... unless it's captured in some dev docs somewhere?

ConorMacBride commented:
Hi @bjlittle, the tests in `tests/subtests/` are very tedious to update (which I realised in #145 after adding only three new tests). All the `tests/subtests/` tests were added in #134 so we can more fully test how the plugin is running internally and verify the outputs under a range of different configurations. In a separate issue/PR this should be discussed/improved, so I would recommend just disabling all the tests in `tests/subtests/` for this PR. I'll think of a way to make them easier to update, and at the very least I'll add more documentation!

bjlittle commented on pytest_mpl/kernels.py:
```diff
@@ -0,0 +1,223 @@
+            value = marker.kwargs.get(self.option)
+            if value is not None:
+                # Override with the decorator marker value.
+                self.hamming_tolerance = int(value)
```
I guess this behaviour is relevant to the discussion over on #154

bjlittle commented:
> Hi @bjlittle, the tests in `tests/subtests/` are very tedious to update (which I realised in #145 after adding only three new tests). All the `tests/subtests/` tests were added in #134 so we can more fully test how the plugin is running internally and verify the outputs under a range of different configurations. In a separate issue/PR this should be discussed/improved, so I would recommend just disabling all the tests in `tests/subtests/` for this PR. I'll think of a way to make them easier to update, and at the very least I'll add more documentation!

Thanks @ConorMacBride :beers: 

I've subsequently raised issue #159 so that we follow through on reinstating the sub-tests... and apologies, I really didn't mean to create extra work for you :fearful:

bjlittle commented on tests/test_pytest_mpl.py:
```diff
@@ -486,3 +490,16 @@ def test_results_always(tmpdir):
+               f"--mpl-hash-library={phash_library}",
+               test_file]
+    code = call_pytest(command)
+    assert code == 0
```
@Cadair @ConorMacBride @dopplershift 

This perceptual hash test is passing across all versions of python, mpl and platforms.

bjlittle commented on tests/subtests/test_subtest.py:
```diff
@@ -4,13 +4,15 @@
 
+# TODO: Remove this skip, see issue https://github.com/matplotlib/pytest-mpl/issues/159
+pytest.skip(reason="temporarily disabled sub-tests", allow_module_level=True)
+
```
@ConorMacBride I'll remove this and rebase when #163 is merged 👍

ConorMacBride replied:
Thanks! Out of curiosity, I tried that and it generates the baseline data correctly. The only modification that's needed to the baseline generation functions is https://github.com/matplotlib/pytest-mpl/pull/163#discussion_r884159274

ConorMacBride reviewed (commented):
This is brilliant! 🎉 I've went through it all and left some comments. Otherwise, I think this feature is ready. Updating the HTML can be a separate PR if that would be easier for reviewing? (As-is the HTML output will work but it won't say which hash kernel is used.) 

Also, (as a separate PR) I think `phash` should be made the default. As it's more lenient to different versions of matplotlib it will make `pytest-mpl` easier to use which would encourage people to use it. (And those who want the stricter SHA hashing can configure that.) As mentioned in other issues, the whole `pytest-mpl` API should be reviewed to decide what other breaking changes should be included in `v1.0.0`.

bjlittle commented on pytest_mpl/kernels.py:
```diff
@@ -0,0 +1,230 @@
+import imagehash
+from PIL import Image
+
+#: The default hamming distance bit tolerance for "similar" imagehash hashes.
+DEFAULT_HAMMING_TOLERANCE = 4
+
+#: The default imagehash hash size (N), resulting in a hash of N**2 bits.
+DEFAULT_HASH_SIZE = 16
+
+#: Level of image detail (high) or structure (low) represented by phash .
+DEFAULT_HIGH_FREQUENCY_FACTOR = 4
```
I'd hope they're reasonable defaults, but it's difficult to judge. They're the defaults we use on SciTools/iris, however I did bump the tolerance from 2 to 4 to get the phash tests to pass on a Red hat Linux distro, rather than Ubuntu.

Typically, I'd be really nervous about raising the default tolerance higher than 4, as it might hide a genuine change that requires manual inspection; given that we're attempting to test images that should pretty much remain invariant.

ConorMacBride replied:
Yeah, it's difficult to understand what effect each of the defaults will have. Not sure how useful this is, but [here's some code](https://gist.github.com/ConorMacBride/a1d8d3239b462bcce3b2fde9b1f24a60) to see what the image looks like during the phash algorithm. A lot of the detail is removed but changes should be detected by subtle differences in the greyscale value. A thought: maybe we should increase the default hash size and the default hamming tolerance so people will have more flexibility when choosing the tolerance? Although if the current values work for SciTools then it's probably safer to use those as a default.

![phash](https://user-images.githubusercontent.com/25007832/170870105-29fa0d9e-40cb-499f-b566-148967e6957d.png)

bjlittle replied:
I guess we can bank what we've got here, as it seems like a reasonable starting point.

Then let the dust settle on this PR, and it'll give you guys time to play and experiment. If there's a concensus, then tweak the defaults based on your experience of how phash behaves.

Does that sound reasonable?

bjlittle commented on setup.cfg:
```diff
@@ -30,6 +30,7 @@ install_requires =
     importlib_resources;python_version<'3.8'
     packaging
     Jinja2
+    imagehash
```
I wasn't sure how you guys dealt with dependencies. I was keeping explicit dependencies to a minimum, and assuming that `Pillow` would always be included through `imagehash` and/or `matplotlib`.

However, your strategy is defo safer 👍

bjlittle commented:
@ConorMacBride Awesome! I'll address your review comments asap. 

But I might take you up on your suggestion for splitting this PR up into a couple of follow-on PRs. Might be a wise strategy 👍

bjlittle commented:
@ConorMacBride Okay, so I've addressed your review comments... and generalised the approach to serializing metadata for the kernel within the hash library and results json. This should extent nicely when we come to add more kernel support for other algorithms.

I guess at this point we're now blocked on this PR until #163 is banked.

That said, if you guys are looking for another `pytest-mpl` core dev, then I'm willing to help out and contribute. I'm completely vested with `pytest-mpl` now, given this PR :+1:

dopplershift commented:
I don't know that we have a procedure in place, but I'm happy to add @bjlittle. I'd appreciate having a :+1: from someone else  though.

ConorMacBride commented:
Hi @bjlittle, do you think you'll have time to complete this PR? I think it just needs the disabled summary tests to be reenabled, branch rebased, and the new configuration documented.

I think it's well tested and the extra summary tests can be added in a separate PR. Full integration with the HTML summary can also be separate (it should work, but will not currently specify the type of hash for example).

bjlittle commented:
Hey @ConorMacBride,

Major apologies for being AWOL. I've been recently maxed out on other project work and just come back for a much needed long summer vacation.

I'm aiming to spin back up on this PR and catch-up on all the recent awesome changes on `pytest-mpl`.

Super keen to get this one banked :+1:

ConorMacBride commented:
Hi @bjlittle, would you like me to rebase this for you? I'm still very keen to get this feature in!

bjlittle commented:
@ConorMacBride Hey, thanks for the nudge, again.

I've been majorly distracted by another major project, but excuses aside let's get this banked 👍

I'm going to work on this next week to get this pr across the line. I may ping you for some support to expedite it, if I get stuck. Is that alright?

bjlittle commented:
I'm closing this PR and replacing it with another
